### PR TITLE
ensure a map passed to strict contains no nil values

### PR DESCRIPTION
**Pull request by brydoncheyney**
Increased levels of intolerance.... StrictMap will ensure that a map passed to the `strict` function contains no nil values.

Example use case: services are often configured at run time using configuration derived from structured data files (edn/yaml). These values are often populated during deployment.  If values are undefined behaviour is not always predictable.  Implementing a nil check ensures a service fails fast when values have been missed.

Interested in feedback on

- technical implementation, obvs
- function naming/conventions
- default behaviour; should nil value check be separated from the current functionality, optional composition, etc.
- should 'nil' be a configurable value? 

also, hi :)

### CODE REVIEW

**Comment by ctford** [comment]
Interesting PR.

A couple of comments from first glance:
* If this were a larger project, folks might object to PRs that aren't targeted. No problem here, but if you upgraded Clojure version and did some other stuff and sometimes that gets PRs rejected.
* On that topic, might it not be a better idea to depend on an older Clojure version? Do we now depend on 1.8? I'm a little fuzzy on how this should work, but perhaps libs should indicate the oldest version they support and let consumers override that in their own projects?
* Do the semantics you're implementing make sense? For example, it is sometimes valid to put nils in a map, if you intentionally do it.
* The change isn't backwards compatible, because if someone is putting nils in a map that would stop working.
* Of course, I don't think this has any significant number of users, so breaking backwards compatibility isn't a problem if it doesn't actually break anyone's code.
* Idea - what if we supported two different kinds of strict map? An "intolerant" map to use your phrase? That way people could opt in to the new behaviour.

In summary though, I think whatever changes to make it still useful for your use case we should do. My replies are a little bit play-acting as if this were a widely deployed project, but I got the feeling from your PR that kind of feedback would be useful to you.

Also, hi. :-)

**Comment by brydoncheyney** [comment]
Excellent points - precisely the response I was looking for, thanks!

- indeed. The concept of a targeted PR makes a lot of sense...
- I'll confess to not really considering the impact of updating dependency versions in a library... I guess the usual workflow of ensuring your project is on the latest dependency versions doesn't really apply to libraries. As it happens, an update of midje was necessary to support map equality tests but, as these dependencies would not be imposed upon projects that use the library, this update would 'exempt' from those rules
- this did give me pause for thought when I pulled this together as, like you say, assigning nil values to a map is a perfectly valid operation - if not my particular use case. I've added the ability to define a predicate instead of a fixed check for nil values.  This also relates somewhat to the next point...
- backwards compatibility. Yep. My bad.
- offering both kinds of strict map would be the best solution, in my opinion...

I've pulled together another commit to reflect the advice above - my only real question is whether the or not the solution is 'idiomatically' the best implementation...  

    ;; constructor (implemented)
    (strict m)        ; strict map, current behaviour
    (strict m pred)   ; less tolerant strict map

vs

    ;; functional composition
    (strict m)    ; strict map, current behaviour
    (def nil-safe (partial restriction pred))
    (def stricter (comp (nil-safe strict))
    (stricter m)  ; less tolerant strict map

the former is simpler, the latter more flexible...

Naturally, there is also the issue of naming... which is hard. Suggestions welcome!

**Comment by ctford** [comment]
I think I prefer the functional composition style. The reason is, I think we're talking about two different kinds of check here:
* One is about *usage* - if I take a valid map and attempt to get something out of it that it doesn't have, don't break it to me gently.
* Another is about *schema* - does this map have any invalid values e.g. `nil` anywhere in its structure.

Within the implementation of the functional composition style, you might be able to do something more elegant using recursion. So for example rather than getting all the nodes out and then checking them (btw, check out `some`), you could recursively construct the stricter map one level at a time.

Thinking about that made me wonder - what are the semantics you want out of nil checking? Should the error occur when you construct the stricter map, or when you try and get the e.g. `nil` out of it?

**Comment by brydoncheyney** [comment]
...as it happens, I did have an implementation using recursion but this didn't allow for easy identification of the path of a failing value, which I considered useful for the exception. Certainly a recursive check may be a more elegant implementation to simply determine if a nil is present.

As for the semantics...  you're right. I'm not entirely sure where this sits within the 'strictly' ns.  The implementation itself blows up on map construction if a predicate is met but it doesn't _enforce_ this if, say, a nil entry is added to a valid map then retrieved.  Indeed, one way to enforce this would be to implement a ...schema[1] :)

It may be that this *isn't* the place to implement such a quick and dirty solution to a relatively specific problem?

Excuse the dumping ground for thoughts but it's been an interesting dialogue!

[1] Given the use case, it seemed overkill to define a schema for every service configuration file but...  this could be done using composition of largely shared elements. For example, creating a service configuration schema from fragments defining database, queue or config.  I guess a quick and simple 'blow up on nils' was the first steps towards a more mature solution...
